apollo-ios/Sources/Apollo/ApolloStore.swift

```diff
@@ -15,13 +13,66 @@ public protocol ApolloStoreSubscriber: AnyObject {
   ///   - store: The store which made the changes
   ///   - changedKeys: The list of changed keys
   ///   - contextIdentifier: [optional] A unique identifier for the request that kicked off this change, to assist in de-duping cache hits for watchers.
+  /// @deprecated
   func store(_ store: ApolloStore,
              didChangeKeys changedKeys: Set<CacheKey>,
              contextIdentifier: UUID?)
+
+  /// A callback that can be received by subscribers for a particular activity case described in `ApolloStore.Activity`
+  ///
+  /// - Parameters:
+  ///   - store: The store which made the changes
+  ///   - activity: The activity that triggered this callback
+  ///   - contextIdentifier: [optional] A unique identifier for the request that kicked off this change, to assist in de-duping cache hits for watchers.
+  func store(_ store: ApolloStore,
+            activity: ApolloStore.Activity,
+            contextIdentifier: UUID?) throws
 }
 
 /// The `ApolloStore` class acts as a local cache for normalized GraphQL results.
 public class ApolloStore {
+  public enum Activity: Hashable {
+    /// Peceived by subscribers BEFORE an action is executed, where the action can be prevented if an error is thrown.
+    /// - Parameters:
+    ///   - perform: The type of action being performed, e.g., load, merge, remove
+    case will(perform: Action)
+
+    /// Received by subscribers AFTER an action has been executed.
+    /// - Parameters:
+    ///   - perform: The type of action that was performed, e.g., load, merge, remove
+    ///   - result: The result of the action, including any relevant data or changed keys
+    case did(perform: Action, outcome: Action.Outcome)
+
+    // Enum to define the types of actions performed in the store
+    public enum Action: Hashable {
+      /// Received by subscribers for records to be loaded from the database for the provided keys.
+      /// - Parameters:
+      ///   - forKeys: The keys that were provided to the store to load records for
+      case loadRecords(forKeys: Set<Apollo.CacheKey>)
+      /// Received by subscribers for records to be merged into the database.
+      /// - Parameters:
+      ///   - records: The records that will be merged into the store
+      case merge(records: RecordSet)
+      /// Received by subscribers for a record to be removed from the database.
+      /// - Parameters:
+      ///   - for: The key for of record that was removed
+      case removeRecord(for: CacheKey)
+      /// Received by subscribers for records matching the provided pattern to be removed from the database.
+      /// - Parameters:
+      ///   - matching: The pattern for whcih matching records were removed
+      case removeRecords(matching: Apollo.CacheKey)
+      /// Received by subscribers for when the database is cleared.
+      case clear
+
+      // Enum to represent the outcome of an action, which can be customized to include more data as needed
+      public enum Outcome: Hashable {
```

Comment:
I'm a bit wary of the outputs being an `enum` vs a return value of a function for similar reasons: it constitutes a breaking change if we ever update the enum while obscuring the output at the same time.

For example, if we had a delegate call like so:

```swift
func store(_ apolloStore: ApolloStore, didLoadRecords forKeys: Set<CacheKey) -> [CacheKey: Record]
``` 

It is much clearer to me what the output is – whereas the alternative is to return an `Output` – which could be one of three things. As a consumer of the API and implementer of a custom `ApolloStoreSubscriber`, it implies that the outputs of these functions aren't mutually exclusive: I could be receiving any one of these three – as opposed to a specific one.

Reply:
I don't think it is supposed to be the return value of the delegate function, rather, just another parameter. But I agree that this doesn't feel right. Specific `Action` types should have specific "outcome" values the provide.

Reply:
Gotcha, I must've misunderstood that – I would think those would be an associated type on the given `Action`s then.